### PR TITLE
MSI 2358: disable in store pickup delivery for multishipping

### PR DESCRIPTION
**Pull request by websnap**
### Description (*)
In-Store Pickup Shipping Method must not be available for Multi-shipping (at least for MVP).
Now it displays like a one more shipping method.
Current pull request will hide it.

### Fixed Issues (if relevant)
1. magento/inventory#2358: Disable In-Store Pickup Delivery for Multishipping

### Manual testing scenarios (*)
1. Put product to the shopping cart
2. Go to shopping cart page
3. Click 'Check Out with Multiple Addresses'
4. Click 'Go to Shipping Information'
5. 'Pick in Store' shipping method is not available in the Shipping Method list

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/InventoryInStorePickupMultishipping/composer.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "magento/module-inventory-in-store-pickup-multishipping",
+    "description": "N/A",
+    "require": {
+        "php": "~7.1.3||~7.2.0",
```
Comment (swnsma) [inline app/code/Magento/InventoryInStorePickupMultishipping/composer.json]: Please add support support of php ~7.3.0

File: app/code/Magento/InventoryInStorePickupMultishipping/Test/Unit/Controller/Checkout/PluginTest.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Tests for plugin which turns Off Multishipping mode for Quote.
+ */
+class PluginTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @var \PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $cartMock;
```
Comment (swnsma) [inline app/code/Magento/InventoryInStorePickupMultishipping/Test/Unit/Controller/Checkout/PluginTest.php]: Please `protected` -> `private`

File: app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryInStorePickupMultishipping\Plugin\Checkout\Controller\Index\Index;
+
+/**
+ * Turns Off Multishipping mode for Quote.
+ */
```
Comment (swnsma) [inline app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php]: Please add description about issue in core and your PR to solve it.

**Review by swnsma — CHANGES_REQUESTED** [review]
Please check my comments.
Please add Integration test.
You can use as the base tests in \Magento\InventoryInStorePickupShipping\Test\Integration\Carrier\InStorePickupTest
But 'in_store_pickup' should not be available if we have 'Multishipping' quote.

UPD
Please fix static tests

**Comment by swnsma** [comment]
@magento run all tests

File: app/code/Magento/InventoryInStorePickupMultishipping/Model/Carrier/Validation/MultiShippingValidator.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function validate(RateRequest $rateRequest): ValidationResult
+    {
+        $errors = [];
+
+        $quote = $this->checkoutSession->getQuote();
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickupMultishipping/Model/Carrier/Validation/MultiShippingValidator.php]: Is the session really the only option? It's highly discouraged to do any session manipulations outside of presentation layer

File: app/code/Magento/InventoryInStorePickupMultishipping/Model/Carrier/Validation/MultiShippingValidator.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+        if ($quote->getIsMultiShipping()) {
+            $errors[] = __('No Pickup Locations available for Multi Shipping');
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickupMultishipping/Model/Carrier/Validation/MultiShippingValidator.php]:
```suggestion
            $errors[] = __('In-Store Pickup is not available with multiple address checkout');
```

I'd look for a more appropriate wording.

File: app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+     *
+     * @param \Magento\Framework\App\Action\Action $subject
+     * @return void
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeExecute(\Magento\Framework\App\Action\Action $subject)
+    {
+        $quote = $this->cart->getQuote();
+        if ($quote->getIsMultiShipping()) {
+            $quote->setIsMultiShipping(0);
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php]: I wouldn't just silently disable multishipping. Maybe we should do other way and remove pickup location assignment if customer enters multishipping with store pickup quote?

Also, it seems like plugin will work regardless of store pickup status, even if it's not used.

File: app/code/Magento/InventoryInStorePickupMultishipping/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# InventoryInStorePickupMultishipping module
+
+The `InventoryInStorePickupMultishipping` module provides business logic for In-Store Pickup functionality on Multishipping Checkout.
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickupMultishipping/README.md:3]: Would make sense to include information on the multishipping disabling in certain conditions.

File: app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryInStorePickupMultishipping\Plugin\Checkout\Controller\Index\Index;
+
+/**
+ * Turns Off Multishipping mode for Quote.
+ */
+class Plugin
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php]: Naming does not provide any context on this plugin

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/inventory?pullRequest=2489) <br/>All committers have signed the CLA.

**Comment by ishakhsuvarov** [comment]
@websnap Thank you for your PR.

Please take a look at review comments and also sign CLA.

File: app/code/Magento/InventoryInStorePickupMultishipping/Model/Carrier/Validation/MultiShippingValidator.php (hunk 4 above)
Comment (swnsma) [reply app/code/Magento/InventoryInStorePickupMultishipping/Model/Carrier/Validation/MultiShippingValidator.php]: Agree.
Need to find the way how to extract quote from RateRequest.

File: app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php (hunk 6 above)
Comment (swnsma) [reply app/code/Magento/InventoryInStorePickupMultishipping/Plugin/Checkout/Controller/Index/Index/Plugin.php]: The reference for the original plugin:
\Magento\Multishipping\Controller\Checkout\Plugin
And PR where it was changed: https://github.com/magento/magento2/pull/24072

**Review by ishakhsuvarov — APPROVED** [review]

**Review by swnsma — APPROVED** [review]

**Pull request merged**